### ISSUE TITLE
Please add helper methods to test contract failures

### ISSUE THREAD

**Issue by digulla** [trufflesuite/truffle#2647]
- [ ] I've asked for help in the [Truffle Gitter](http://gitter.im/Consensys/truffle) before filing this issue.

---------------------------

I'd like to make sure contracts contain the necessary `require()` calls and that they trigger under the correct circumstances.

This is especially important when using modifiers and inheritance between contracts.

    contract Foo {
        address owner;

        constructor() public {
            owner = msg.sender;
        }

        function transferOwnership(address newOwner) public {
            require(owner == msg.sender, 'Only the current owner can transfer ownership');
        }
    }

Proposed syntax:

    contract('Foo Tests', (accounts) => {
        const deployer = accounts[0]
        const someoneElse = accounts[1]

        it('should be possible to test require()', async () => {
            const foo = Foo.new() // implicit: from: deployer

            // Try to steal foo
            expect(() => 
                foo.transferOwnership(someoneElse, {from: someoneElse})
            ).to.require('Only the current owner can transfer ownership')
        })
    })

I tried to make this work with this code:

    // Based on https://stackoverflow.com/a/56104568/34088
    const getErrorTypeName = (errorType) => {
        let result = type(errorType);
        if (result === 'function') {
            result = errorType.name;
        }
        return result;
    }

    const expectThrowsAsync = async (method, errorType, errorMessage) => {
        let error = null;
        try {
            await method();
        } catch (err) {
            error = err
        }
        if (error) {
            if (typeof(errorType) === 'string') {
                expect(getErrorTypeName(error)).to.equal(errorType);
            } else {
                expect(error).to.be.an.instanceof(errorType);
            }
        } else {
            const typeName = getErrorTypeName(errorType);
            assert.fail(`Missing exception of type "${typeName}"`);
        }
        if (errorMessage) {
            expect(error.message).to.equal(errorMessage);
        }
    }

but this **sometimes** fails when there are more `contract()`s in a single test. If I use this in a method of the first `contract()`, then I'll get weird exceptions ("Returned values aren't valid, did it run Out of Gas?" or "Uncaught Error: abort({}). Build with -s ASSERTIONS=1 for more info." or the tests pass when they shouldn't. I don't know enough about mocha/chai/truffle to understand why this happens.

Here is the full code for `foo_test.ts`:

```
const Foo = artifacts.require("Foo");

var type = require('type-detect');

// Based on https://stackoverflow.com/a/56104568/34088
const getErrorTypeName = (errorType) => {
    let result = type(errorType);
    if (result === 'function') {
        result = errorType.name;
    }
    return result;
}

const expectThrowsAsync = async (method, errorType, errorMessage) => {
    let error = null;
    try {
        await method();
    } catch (err) {
        error = err
    }
    if (error) {
        if (typeof(errorType) === 'string') {
            expect(getErrorTypeName(error)).to.equal(errorType);
        } else {
            expect(error).to.be.an.instanceof(errorType);
        }
    } else {
        const typeName = getErrorTypeName(errorType);
        assert.fail(`Missing exception of type "${typeName}"`);
    }
    if (errorMessage) {
        expect(error.message).to.equal(errorMessage);
    }
}

contract('Foo Tests', (accounts) => {
    const deployer = accounts[0]
    const newOwner = accounts[1]
    const someoneElse = accounts[2]

    it('should work', async () => {
        const foo = await Foo.new()

        await foo.transferOwnership(newOwner)
    })

    it('stealing should not work', async () => {
        const foo = await Foo.new()

        expectThrowsAsync(() =>
            foo.transferOwnership(someoneElse, {from: someoneElse}),
            Error,
            'xxx' // Should fail because of mismatch
        )
    })
})

contract('Second Test', (accounts) => {
    it('should always pass', () => {
        expect(true).to.be.true
    })
})
```

My `package.json`:

```
{
    "name": "ballot",
    "version": "0.0.1",
    "dependencies": {
        "truffle": "^5.1.1"
    },
    "devDependencies": {
        "chai": "^4.2.0",
        "solidity-coverage": "^0.6.7",
        "ts-node": "^8.5.2",
        "typescript": "^3.7.2"
    }
}
```

## Environment

* Operating System: 
* Ethereum client:
* Truffle version (`truffle version`):
* node version (`node --version`):
* npm version (`npm --version`):

**Comment by digulla** [comment]
Related to #2497 

Note: The tests above will only start to pass when after adding

    if (typeof(err) === 'string') {
      err = Error(err);
    }

after the line

    Runner.prototype.uncaught = function(err) {

of `./node_modules/truffle/node_modules/mocha/lib/runner.js` (around line 730).

**Comment by gnidan** [comment]
This is a neat idea, thanks for suggesting it!

You might be interested to take a look at @rkalis's https://github.com/rkalis/truffle-assertions library.

**Comment by digulla** [comment]
I just found `truffle-assertions`. It also supports testing events. It would be great if you could integrate both features into the core testing framework.

**Comment by rkalis** [comment]
I'm not opposed to contributing the truffle-assertions functionality into the Truffle codebase some time down the line. Can't really make it a priority though, as I think the current workflow (`npm i -D truffle-assertions` and import it in the test files) is good enough for most people, and reduces the maintenance burden on the core Truffle team.

**Comment by digulla** [comment]
Agreed. `truffle-assertions` works me so I'd vote for priority "low" / "minor". Thanks :-)
